factfinder/main.py

```diff
@@ -58,7 +79,91 @@ def median_design_factor(self, pff_variable) -> float:
         """
         return self.median[pff_variable]["design_factor"]
 
-    def calculate_median_variable(self, pff_variable, geotype) -> pd.DataFrame:
+    def calculate(self, pff_variable: str, geotype: str) -> pd.DataFrame:
+        """
+        Main user interface, going through different following steps:
+        1. calculate_c_e_m_p_z
+        2. rounding
+        3. assigning geoid
+        """
+        # 0. Initialize Variable class instance
+        v = self.create_variable(pff_variable)
+        # 1. get calculated values (c,e,m,p,z)
+        df = self.calculate_c_e_m_p_z(v, geotype)
+        # 2. rounding
+        df = self.rounding(df, v.rounding)
+        # 3. last round of data cleaning
+        df = self.cleaning(df, v, geotype)
+        return df
+
+    def rounding(self, df: pd.DataFrame, digits: int) -> pd.DataFrame:
+        """
+        Round c, e, m, p, z fields based on rounding digits from metadata
+        """
+        df["c"] = df["c"].round(1)
+        df["e"] = df["e"].round(digits)
+        df["m"] = df["m"].round(digits)
+        df["p"] = df["p"].round(1)
+        df["z"] = df["z"].round(1)
+        return df
+
+    def cleaning(self, df: pd.DataFrame, v: Variable, geotype: str) -> pd.DataFrame:
+        """
+        last step data cleaning based on rules below:
+        """
+        # negative values are invalid
+        df.loc[df.c < 0, "c"] = np.nan
+        df.loc[df.e < 0, "e"] = np.nan
+        df.loc[df.m < 0, "m"] = np.nan
+        df.loc[df.p < 0, "p"] = np.nan
+        df.loc[df.z < 0, "z"] = np.nan
+
+        # p has to be less or equal to 100
+        df.loc[df.p > 100, "p"] = np.nan
+
+        # If p = np.nan/, then z = np.nan
+        df.loc[(df.p.isna()) | (df.p == 100), "z"] = np.nan
+
+        df.loc[df.e == 0, "c"] = np.nan
+        df.loc[
+            df.e == 0 & 
+            ~df.pff_variable.isin(self.base_variables), "m"
+        ] = np.nan
+        df.loc[
+            df.e == 0 & 
+            df.pff_variable.isin(self.base_variables) & 
+            df.m.isna(), "m"
+        ] = 0
+        df.loc[df.e == 0, "p"] = np.nan
```

Comment:
Again, why? If p represents a proportion estimate, and there the count of something is estimated as 0, wouldn't the proportion also be estimated as 0?